File: Chapter9_Example/src/main/java/BusSpy.java
import org.assertj.core.api.Assertions;

import java.util.ArrayList;
import java.util.List;

public class BusSpy implements IBus {
    private final List<String> _sentMessages = new ArrayList<>();

    public void send(String message) {
        _sentMessages.add(message);
    }

    public BusSpy shouldSendNumberOfMessages(int number) {
        Assertions.assertThat(number).isEqualTo(_sentMessages.size());
        return this;
    }

    public BusSpy withEmailChangedMessage(int userId, String newEmail) {
        String message = "Type : USER EMAIL CHANGED; Id : " + userId + "; NewEmail : " + newEmail;
        Assertions.assertThat(_sentMessages).contains(message);
        return this;
    }
}

File: unit-testing/src/test/kotlin/book/chapter9/v2/BusSpy.kt
package book.chapter9.v2

import book.chapter9.v1.IBus
import org.junit.jupiter.api.Assumptions

class BusSpy: IBus {
    private val sentMessages = mutableListOf<String>()

    override fun send(message: String) {
        sentMessages.add(message)
    }

    fun shouldSendNumberOfMessages(number: Int): BusSpy {
        Assumptions.assumeTrue(number == sentMessages.size)

        return this
    }

    fun withEmailChangedMessage(userId: Int, newEmail: String): BusSpy {
        val message = "Type: USER EMAIL CHANGED; " +
                "Id: $userId; " +
                "NewEmail: $newEmail"

        Assumptions.assumeTrue(sentMessages.contains(message))

        return this
    }
}